feat: support derived source for knn with other fields - porting (#488)
opensearch-project/opensearch-jvector: docs/derived_source.md

### OpenSearch 3.7 and later

Both options are available:

- `index.knn.derived_source.enabled`
- `index.derived_source.enabled`

If both are provided, `index.derived_source.enabled` takes precedence.

For KNN-only use cases, `index.knn.derived_source.enabled` remains valid. For indexes that should use derived source across all supported fields, prefer `index.derived_source.enabled`.

---

## Setting Behavior

General rules:

- Indexes created with `index.knn.derived_source.enabled` continue to work after upgrade
- The only supported migration path between implementations is to reindex data
- If both settings are present on OpenSearch 3.7+, core derived source takes precedence
- If the core setting is not enabled (default), `opensearch-jvector` can continue handling derived source for vector fields

This preserves compatibility for existing `opensearch-jvector` indexes while allowing newer clusters to adopt the core OpenSearch implementation. By default `index.derived_source.enabled` is disabled.

---

## Requirements and Limitations

`opensearch-jvector` plugin implementation

Requirements:

- `index.knn.derived_source.enabled` requires `index.knn: true`

> - `_source` must remain enabled for the jVector plugin implementation

Reviewer: ```suggestion
- `_source` must remain enabled for the `opensearch-jvector` plugin implementation
```

Supported Features:

- Nested vector fields
- Works independently of other field types

Limitations:
- copy_to not supported
- `opensearch-jvector`-derived source is disabled when segment replication with local node-to-node replication is enabled

OpenSearch Core Approach

Requirements:

> - `_source` must remain enabled for the jVector plugin implementation

Reviewer: ```suggestion
- `_source` must remain enabled for the `opensearch-jvector` plugin implementation
```

# Derived Source

## Overview

> jVector supports derived source for KNN vector fields through the plugin setting `index.knn.derived_source.enabled`.

Reviewer: ```suggestion
`opensearch-jvector` supports derived source for KNN vector fields through the plugin setting `index.knn.derived_source.enabled`.
```

> Starting from OpenSearch 3.7, you can also use the OpenSearch core derived source setting `index.derived_source.enabled`. This adds support for derived source across all supported fields, including `knn_vector` fields when used with jVector.

Reviewer: ```suggestion
Starting from OpenSearch 3.7, you can also use the OpenSearch core derived source setting `index.derived_source.enabled`. This adds support for derived source across all supported fields, including `knn_vector` fields when used with `opensearch-jvector`.
```

Derived source reduces stored `_source` overhead by reconstructing fields when documents are read back. This is useful when you want to reduce storage overhead while still retrieving documents that include vectors and regular fields from the same index.

---

## Configuration Options

> There are two derived source mechanisms relevant to jVector indexes:

Reviewer: ```suggestion
There are two derived source mechanisms relevant to k-NN indexes:
```

- OpenSearch core derived source via `index.derived_source.enabled`

> - jVector plugin derived source via `index.knn.derived_source.enabled`

Reviewer: ```suggestion
- `opensearch-jvector` plugin derived source via `index.knn.derived_source.enabled`
```

> ### jVector plugin derived source

Reviewer: ```suggestion
### `opensearch-jvector` plugin derived source
```

> This mechanism is available independently of OpenSearch 3.7 and is controlled by the plugin setting:

Reviewer 2: ```suggestion
This mechanism is similar to k-nn plugin, and is controlled by the plugin setting:
```

```json
{
  "settings": {
    "index.knn": true,
    "index.knn.derived_source.enabled": true
  }
}
```

> This option is specific to jVector/KNN behavior and applies only when KNN is enabled on the index.

Reviewer: ```suggestion
This option is specific to `opensearch-jvector` behavior and applies only when KNN index setting is enabled on the index.
```

### OpenSearch core derived source

Starting from OpenSearch 3.7, you can use the core setting:

```json
{
  "settings": {
    "index.knn": true,
    "index.derived_source.enabled": true
  }
}
```

> When `index.derived_source.enabled` is enabled, the core implementation takes precedence over the jVector-specific implementation, including for KNN fields.

Reviewer: ```suggestion
When `index.derived_source.enabled` is enabled, the core implementation takes precedence over the `opensearch-jvector`-specific implementation, including for KNN fields.
```

---

## Version Behavior

### Before OpenSearch 3.7

> Use the jVector plugin setting:

Reviewer: ```suggestion
Use the `opensearch-jvector` plugin setting:
```

> ### Before OpenSearch 3.7

Reviewer 2: ```suggestion
### Before OpenSearch JVector 3.7
```

Reviewer 2: I think you mean the plugin release 3.7, where this feature is going to be supported. 

Not to be confused with opensearch 3.7 release (as opensearch announced this feature last year). Could you do a sanity check on the rest of the documentation, to make sure we're referring to plugin 3.7 release

Author: I mean actually both, since core solution was added in 3.7, and property setting is from core, not plugin

- `index.knn.derived_source.enabled`

> The core setting `index.derived_source.enabled` does not apply to this jVector use case before OpenSearch 3.7.

Reviewer 2: ```suggestion
The core setting `index.derived_source.enabled` does not apply to this jVector plugin use case before OpenSearch 3.7.
```

### OpenSearch 3.7 and later

Both options are available:

- `index.knn.derived_source.enabled`
- `index.derived_source.enabled`

If both are provided, `index.derived_source.enabled` takes precedence.

For KNN-only use cases, `index.knn.derived_source.enabled` remains valid. For indexes that should use derived source across all supported fields, prefer `index.derived_source.enabled`.

---

## Setting Behavior

General rules:

- Indexes created with `index.knn.derived_source.enabled` continue to work after upgrade
- The only supported migration path between implementations is to reindex data
- If both settings are present on OpenSearch 3.7+, core derived source takes precedence

> - If the core setting is not enabled, jVector can continue handling derived source for vector fields

Reviewer: ```suggestion
- If the core setting is not enabled, `opensearch-jvector` can continue handling derived source for vector fields
```

> This preserves compatibility for existing jVector indexes while allowing newer clusters to adopt the core OpenSearch implementation.

Reviewer: ```suggestion
This preserves compatibility for existing `opensearch-jvector` indexes while allowing newer clusters to adopt the core OpenSearch implementation.
```

---

## Requirements and Limitations

> jVector plugin implementation

Reviewer: ```suggestion
`opensearch-jvector` plugin implementation
```

> 

Reviewer 2: What is the default behavior for `index.derived_source.enabled`? 

For e.g., any k-nn index created by a user will set the `index.knn.derived_source.enabled` as true.  
https://github.com/opensearch-project/opensearch-jvector/blob/b74bcc3f85a78c5fb12a8fa1eb4d98877eebf850/src/main/java/org/opensearch/knn/plugin/JVectorKNNPlugin.java#L223

Requirements:

- `index.knn.derived_source.enabled` requires `index.knn: true`
- `_source` must remain enabled for the jVector plugin implementation

Supported Features:

- Nested vector fields
- Works independently of other field types

> - Field-level enablement

Reviewer 2: @tomkdgun , @reta - is 'field-level enablement' accurate here for the knn version of derived sources? The setting is done at the index level as you show in the example below, so not sure I follow how this is field level.

Author: Solution is similar as on opensearch-jvector, index level. I was mean something else, but agree this is misleading, "Works independently of other field types" is sufficient

Limitations:
- copy_to not supported

> - jVector-derived source is disabled when segment replication with local node-to-node replication is enabled

Reviewer: ```suggestion
- `opensearch-jvector`-derived source is disabled when segment replication with local node-to-node replication is enabled
```

OpenSearch Core Approach

Requirements:

- `_source` must remain enabled for the jVector plugin implementation

Supported Features:

> - Eliminate redundant storage of the entire _source field by reconstructing it from indexed data structures (doc values/stored fields)

Reviewer: ```suggestion
- Eliminate redundant storage of the entire `_source` field by reconstructing it from indexed data structures (doc values/stored fields)
```

- Unified approach for all field types
- Translog-based real-time reads with consistent formatting
- Index-level enablement

Limitations:

- Nested objects not supported
- All-or-nothing: If ANY field doesn't support derived source, entire feature fails for the index (normalizer, geo_shape, date_range, copy_to not supported)
- Field-level enablement/disablement not supported (doc_values=false, store=false not supported)

For core-derived source limitations, see the OpenSearch documentation:
https://docs.opensearch.org/latest/mappings/metadata-fields/source/#limitations

---

## Example Index Creation

> ### Using jVector plugin derived source

Reviewer: ```suggestion
### Using `opensearch-jvector` plugin derived source
```

```bash
curl -X PUT "http://localhost:9200/my-derived-source-index" \
  -H "Content-Type: application/json" -d '
{
  "settings": {
    "index.knn": true,
    "index.knn.derived_source.enabled": true
  },
  "mappings": {
    "properties": {
      "title": { "type": "text" },
      "category": { "type": "keyword" },
      "embedding": {
        "type": "knn_vector",
        "dimension": 128,
        "method": {
          "name": "disk_ann",
          "engine": "jvector",
          "space_type": "cosinesimil"
        }
      }
    }
  }
}'
```

### Using OpenSearch core derived source

```bash
curl -X PUT "http://localhost:9200/my-derived-source-index" \
  -H "Content-Type: application/json" -d '
{
  "settings": {
    "index.knn": true,
    "index.derived_source.enabled": true
  },
  "mappings": {
    "properties": {
      "title": { "type": "text" },
      "category": { "type": "keyword" },
      "embedding": {
        "type": "knn_vector",
        "dimension": 128,
        "method": {
          "name": "disk_ann",
          "engine": "jvector",
          "space_type": "cosinesimil"
        }
      }
    }
  }
}'
```

---

## Summary

> Use `index.knn.derived_source.enabled` for the jVector plugin implementation. Starting from OpenSearch 3.7, you can also use `index.derived_source.enabled` for the core implementation, which takes precedence when both settings are present.

Reviewer: ```suggestion
Use `index.knn.derived_source.enabled` for the `opensearch-jvector` plugin implementation or when support for nested fields is required. Starting from OpenSearch 3.7, you can also use `index.derived_source.enabled` for the core implementation, which takes precedence when both settings are present.
```

Author: @reta Comments addressed
